File: 0506-relative-ranks/0506-relative-ranks.py
class Solution:
    def findRelativeRanks(self, score: List[int]) -> List[str]:
        n = len(score)

        maxScore = max(score)

        score2Index = [0] * (maxScore + 1)

        for i in range(n):
            score2Index[score[i]] = i + 1

        ans = [""] * n

        place = 1

        for i in range(maxScore, -1, -1):

            if score2Index[i] == 0:
                continue

            actualIndex = score2Index[i] - 1
            if place == 1:
                ans[actualIndex] = "Gold Medal"
            elif place == 2:
                ans[actualIndex] = "Silver Medal"
            elif place == 3:
                ans[actualIndex] = "Bronze Medal"
            else:
                ans[actualIndex] = str(place)
            place += 1

        return ans

File: 0506-relative-ranks/0506-relative-ranks.cpp
class Solution {
public:
    std::vector<std::string> findRelativeRanks(std::vector<int>& score) {
        int n = score.size();

        int maxScore = 0;
        for (int i = 0; i < n; i++) {
            maxScore = std::max(maxScore, score[i]);
        }

        std::vector<int> score2Index(maxScore + 1);

        for (int i = 0; i < n; i++) {
            score2Index[score[i]] = i + 1;
        }

        std::vector<std::string> ans(n);

        int place = 1;

        for (int i = maxScore; i >= 0; i--) {

            if (score2Index[i] == 0) continue;

            int actualIndex = score2Index[i] - 1;
            if (place == 1) {
                ans[actualIndex] = "Gold Medal";
            }
            else if (place == 2) {
                ans[actualIndex] = "Silver Medal";
            }
            else if (place == 3) {
                ans[actualIndex] = "Bronze Medal";
            }
            else {
                ans[actualIndex] = std::to_string(place);
            }
            place++;
        }

        return ans;
    }
};